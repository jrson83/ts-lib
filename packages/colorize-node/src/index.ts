import { inspect } from 'node:util'

const colors = inspect.colors
const stdout = process.stdout

const noColorSupport = stdout.isTTY && !stdout.hasColors()

// biome-ignore lint/suspicious/noConstEnum: removed during compilation
const enum Colors {
  reset,
  bold,
  dim,
  italic,
  underline,
  blink,
  inverse,
  hidden,
  strikethrough,
  doubleunderline,
  black,
  red,
  green,
  yellow,
  blue,
  magenta,
  cyan,
  white,
  bgBlack,
  bgRed,
  bgGreen,
  bgYellow,
  bgBlue,
  bgMagenta,
  bgCyan,
  bgWhite,
  framed,
  overlined,
  gray,
  redBright,
  greenBright,
  yellowBright,
  blueBright,
  magentaBright,
  cyanBright,
  whiteBright,
  bgGray,
  bgRedBright,
  bgGreenBright,
  bgYellowBright,
  bgBlueBright,
  bgMagentaBright,
  bgCyanBright,
  bgWhiteBright,
}

type Colorize = {
  [k in keyof typeof Colors]: (
    input: string | number | null | undefined
  ) => string
}

const colorize = Object.fromEntries(
  Object.entries(colors).map(([color, code]) => [
    color,
    (str) => {
      if (noColorSupport || !code) return str
      return `\x1b[${code[0]}m${str}\x1b[${code[1]}m`
    },
  ])
) as Colorize

export { colorize }
