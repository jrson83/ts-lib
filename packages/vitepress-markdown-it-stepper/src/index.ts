import type MarkdownIt from 'markdown-it'
import container from 'markdown-it-container'
import type { RenderRule } from 'markdown-it/lib/renderer.mjs'

export interface MarkdownItStepperOptions {
  /** The wrapper Element `tagName`. defaults to `div` */
  wrapperTagName?: string

  /** The wrapper Element `class`. defaults to `stepper` */
  wrapperClassName?: string

  /** The title Element `tagName`. defaults to `span` */
  titleTagName?: string

  /** The title Element `class`. defaults to `stepper-title` */
  titleClassName?: string

  /** The content Element `tagName`. defaults to `div` */
  contentTagName?: string

  /** The content Element `class`. defaults to `stepper-content` */
  contentClassName?: string
}

function markdownItStepper(
  md: MarkdownIt,
  options: MarkdownItStepperOptions = {}
): void {
  const opts: MarkdownItStepperOptions = {
    wrapperTagName: 'div',
    wrapperClassName: 'stepper',
    titleTagName: 'span',
    titleClassName: 'stepper-title',
    contentTagName: 'div',
    contentClassName: 'stepper-content',
    ...options,
  }

  createContainer(md, opts)
}

function createContainer(md: MarkdownIt, options: MarkdownItStepperOptions) {
  md.use(container, 'stepper', {
    validate(params) {
      return !!params.trim().match(/^stepper\s*(.*)$/)
    },
    render(tokens, idx) {
      const m = tokens[idx]?.info.trim().match(/^stepper\s*(.*)$/)
      if (tokens[idx]?.nesting === 1) {
        const title = m && m.length > 1 ? m[1] : undefined
        return (
          `<${options.wrapperTagName} class='${options.wrapperClassName}'>\n` +
          (title
            ? `<${options.titleTagName} class='${options.titleClassName}'>${title}</${options.titleTagName}>\n`
            : '') +
          `<${options.contentTagName} class='${options.contentClassName}'>\n`
        )
      } else {
        return (
          `</${options.contentTagName}>\n` + `</${options.wrapperTagName}>\n`
        )
      }
    },
  } satisfies { render: RenderRule; validate: (params: string) => boolean })
}

export { markdownItStepper }
