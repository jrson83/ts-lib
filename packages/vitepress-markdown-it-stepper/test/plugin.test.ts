import { readFileSync } from 'node:fs'
import { resolve } from 'node:path'
import MarkdownIt from 'markdown-it'
import { describe, expect, it } from 'vitest'

import { type MarkdownItStepperOptions, markdownItStepper } from '../src'

const cwd = process.cwd()
const resolvePath = (file: string) =>
  resolve(
    cwd,
    'packages',
    'vitepress-markdown-it-stepper',
    'test',
    'fixtures',
    file
  )

function createProcessor(options?: MarkdownItStepperOptions) {
  return MarkdownIt().use<MarkdownItStepperOptions>(markdownItStepper, options)
}

describe('vitepressPluginMarkdownItStepper', () => {
  it('renders empty string', () => {
    const data = createProcessor().render(
      readFileSync(resolvePath('empty.md'), { encoding: 'utf8' })
    )
    expect(data).toMatchSnapshot()
  })

  it('renders single container with title & custom options', () => {
    const data = createProcessor({
      wrapperTagName: 'section',
      wrapperClassName: 'my-wrapper-class',
      titleTagName: 'div',
      titleClassName: 'my-title-class',
    }).render(
      readFileSync(resolvePath('single-title.md'), { encoding: 'utf8' })
    )
    expect(data).toMatchSnapshot()
  })

  it('renders single container without title', () => {
    const data = createProcessor().render(
      readFileSync(resolvePath('single.md'), { encoding: 'utf8' })
    )
    expect(data).toMatchSnapshot()
  })

  it('renders single container with title', () => {
    const data = createProcessor().render(
      readFileSync(resolvePath('single-title.md'), { encoding: 'utf8' })
    )
    expect(data).toMatchSnapshot()
  })

  it('renders multiple containers without titles', () => {
    const data = createProcessor().render(
      readFileSync(resolvePath('multiple.md'), { encoding: 'utf8' })
    )
    expect(data).toMatchSnapshot()
  })

  it('renders multiple containers with titles', () => {
    const data = createProcessor().render(
      readFileSync(resolvePath('multiple-title.md'), { encoding: 'utf8' })
    )
    expect(data).toMatchSnapshot()
  })

  it('renders multiple containers with & without titles', () => {
    const data = createProcessor().render(
      readFileSync(resolvePath('multiple-both.md'), { encoding: 'utf8' })
    )
    expect(data).toMatchSnapshot()
  })

  it('renders nested containers without titles', () => {
    const data = createProcessor().render(
      readFileSync(resolvePath('nested.md'), { encoding: 'utf8' })
    )
    expect(data).toMatchSnapshot()
  })

  it('renders nested containers with titles', () => {
    const data = createProcessor().render(
      readFileSync(resolvePath('nested-title.md'), { encoding: 'utf8' })
    )
    expect(data).toMatchSnapshot()
  })
})
