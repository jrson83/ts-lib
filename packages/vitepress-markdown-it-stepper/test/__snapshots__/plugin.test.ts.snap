// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`vitepressPluginMarkdownItStepper > renders empty string 1`] = `""`;

exports[`vitepressPluginMarkdownItStepper > renders multiple containers with & without titles 1`] = `
"<h1>Heading</h1>
<div class='stepper'>
<div class='stepper-content'>
<p>Paragraph</p>
<ul>
<li>List item #1</li>
<li>List item #2</li>
</ul>
</div>
</div>
<div class='stepper'>
<span class='stepper-title'>This is a title</span>
<div class='stepper-content'>
<p>Another paragraph</p>
<ul>
<li>Another list item #1</li>
<li>Another list item #2</li>
</ul>
</div>
</div>
"
`;

exports[`vitepressPluginMarkdownItStepper > renders multiple containers with titles 1`] = `
"<h1>Heading</h1>
<div class='stepper'>
<span class='stepper-title'>This is a title</span>
<div class='stepper-content'>
<p>Paragraph</p>
<ul>
<li>List item #1</li>
<li>List item #2</li>
</ul>
</div>
</div>
<div class='stepper'>
<span class='stepper-title'>This is another title</span>
<div class='stepper-content'>
<p>Another paragraph</p>
<ul>
<li>Another list item #1</li>
<li>Another list item #2</li>
</ul>
</div>
</div>
"
`;

exports[`vitepressPluginMarkdownItStepper > renders multiple containers without titles 1`] = `
"<h1>Heading</h1>
<div class='stepper'>
<div class='stepper-content'>
<p>Paragraph</p>
<ul>
<li>List item #1</li>
<li>List item #2</li>
</ul>
</div>
</div>
<div class='stepper'>
<div class='stepper-content'>
<p>Another paragraph</p>
<ul>
<li>Another list item #1</li>
<li>Another list item #2</li>
</ul>
</div>
</div>
"
`;

exports[`vitepressPluginMarkdownItStepper > renders nested containers with titles 1`] = `
"<h1>Heading</h1>
<div class='stepper'>
<span class='stepper-title'>This is a title</span>
<div class='stepper-content'>
<p>Paragraph</p>
<ul>
<li>List item #1</li>
<li>List item #2</li>
</ul>
<div class='stepper'>
<span class='stepper-title'>This is another title</span>
<div class='stepper-content'>
<p>Nested paragraph</p>
<ul>
<li>Nested list item #1</li>
<li>Nested list item #2</li>
</ul>
</div>
</div>
</div>
</div>
"
`;

exports[`vitepressPluginMarkdownItStepper > renders nested containers without titles 1`] = `
"<h1>Heading</h1>
<div class='stepper'>
<div class='stepper-content'>
<p>Paragraph</p>
<ul>
<li>List item #1</li>
<li>List item #2</li>
</ul>
<div class='stepper'>
<div class='stepper-content'>
<p>Nested paragraph</p>
<ul>
<li>Nested list item #1</li>
<li>Nested list item #2</li>
</ul>
</div>
</div>
</div>
</div>
"
`;

exports[`vitepressPluginMarkdownItStepper > renders single container with title & custom options 1`] = `
"<h1>Heading</h1>
<section class='my-wrapper-class'>
<div class='my-title-class'>This is a title</div>
<div class='stepper-content'>
<p>Paragraph</p>
<ul>
<li>List item #1</li>
<li>List item #2</li>
</ul>
</div>
</section>
"
`;

exports[`vitepressPluginMarkdownItStepper > renders single container with title 1`] = `
"<h1>Heading</h1>
<div class='stepper'>
<span class='stepper-title'>This is a title</span>
<div class='stepper-content'>
<p>Paragraph</p>
<ul>
<li>List item #1</li>
<li>List item #2</li>
</ul>
</div>
</div>
"
`;

exports[`vitepressPluginMarkdownItStepper > renders single container without title 1`] = `
"<h1>Heading</h1>
<div class='stepper'>
<div class='stepper-content'>
<p>Paragraph</p>
<ul>
<li>List item #1</li>
<li>List item #2</li>
</ul>
</div>
</div>
"
`;
